### PR TITLE
Intel and Scouting System

### LINKED ISSUE #85: [feature] World Scouting
This is the active part of #84. The primary goal here is to gain visibility to rooms to trigger the room intel process.

### PR DESCRIPTION
**Pull request by tedivm**
This gives everything needed for #84 except the portal data. It resolves #85.

This provides two new systems-

## Intel

The intel system has a few components-

* New `Room` functions- `Room.prototype.getIntel`, `Room.getIntel`, `Room.prototype.saveIntel`, `Room.getResourcesPositions`, `Room. requestIntel`, and `Room. getScoutTarget`. These functions are backed by a segment but use a memory buffer to reduce CPU usage when saving.

* Scouting targets are set automatically when calls to `getIntel` fail to retrieve data or gets data that is too old. The `Room.requestIntel` function can be called to manually add a target.

* If no valid targets are available then `Room. getScoutTarget` will return a neighboring room. It will prioritize unseen rooms first (if it encounters multiple unseen rooms it picks the closest to the supplied creep's parent room), and then will prioritize the room that hasn't been updated recently.

## Scouting

A new creep role, `Spook`, has been added. It does a few things-

* If it sees hostile construction sites in a room it steps on them.

* If it sees an controller that isn't signed by the creep's player then the creep will sign it.

* If it has nothing else to do it requests a new room target from `Room.getScoutTarget` and then travels to that room.

### CODE REVIEW

File: src/extends/room/intel.js [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+      roominfo = {}
+    }
+  }
+
+  roominfo[INTEL_UPDATED] = Game.time
+
+  // Record room owner, level.
+  if (this.controller) {
+    if (this.controller.owner) {
+      roominfo[INTEL_OWNER] = this.controller.owner
```
Comment (TheRealMaxion) [inline src/extends/room/intel.js]: shouldn't this be saving owner.username?

File: src/roles/spook.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        }
+      }
+    }
+
+    // Move on to the next room
+    this.migrateRooms(creep)
+  }
+
+  stompConstruction (creep) {
+    const construction = creep.pos.findClosestByRange(FIND_HOSTILE_CONSTRUCTION_SITES)
```
Comment (TheRealMaxion) [inline src/roles/spook.js]: if we don't filter out construction sites in the same position as the creep then the creep can get stuck on sites that are instantly replaced

Comment (tedivm) [reply src/roles/spook.js]: I thought you couldn't place a construction site where a hostile creep was standing?

Comment (TheRealMaxion) [reply src/roles/spook.js]: I've had stomping creeps get stuck that way before at least.

**Pull request merged**